espnet2/hubert/espnet_model.py

```diff
@@ -0,0 +1,314 @@
+            y_pad: torch.Tensor, y_pad_length: torch.Tensor,
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Frontend + Encoder. Note that this method is used by asr_inference.py
+
+        Args:
+            speech: (Batch, Length, ...)
+            speech_lengths: (Batch, )
+        """
+        with autocast(False):
+            # 1. Extract feats
+            feats, feats_lengths = self._extract_feats(speech, speech_lengths)
+
+            # 2. Data augmentation
+            if self.specaug is not None and self.training:
+                feats, feats_lengths = self.specaug(feats, feats_lengths)
+
+            # 3. Normalization for feature: e.g. Global-CMVN, Utterance-CMVN
+            if self.normalize is not None:
+                feats, feats_lengths = self.normalize(feats, feats_lengths)
+
+        # Pre-encoder, e.g. used for raw input data
+        if self.preencoder is not None:
+            feats, feats_lengths = self.preencoder(feats, feats_lengths)
+
+        # 4. Forward encoder
+        # feats: (Batch, Length, Dim)
+        # -> encoder_out: (Batch, Length2, Dim2)
+        encoder_out = self.encoder(
+            feats, feats_lengths, y_pad, y_pad_length
+        )
+
+        if hasattr(self.encoder, "encoder"):
+            logp_m_list = self.encoder.encoder.get_logits(encoder_out, True)
+            assert self.pred_masked_weight == 0 or len(logp_m_list) > 0
+        
+            logp_u_list = self.encoder.encoder.get_logits(encoder_out, False)
+            assert self.pred_nomask_weight == 0 or len(logp_u_list) > 0
+        else:
+            # TODO : fix raising error
+            raise ValueError
+
+        return encoder_out
+
+    def _extract_feats(
+        self, speech: torch.Tensor, speech_lengths: torch.Tensor
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        assert speech_lengths.dim() == 1, speech_lengths.shape
+
+        # for data-parallel
+        speech = speech[:, : speech_lengths.max()]
+
+        if self.frontend is not None:
+            # Frontend
+            #  e.g. STFT and Feature extract
+            #       data_loader may send time-domain signal in this case
+            # speech (Batch, NSamples) -> feats: (Batch, NFrames, Dim)
+            feats, feats_lengths = self.frontend(speech, speech_lengths)
+        else:
+            # No frontend and no feature extract
+            feats, feats_lengths = speech, speech_lengths
+        return feats, feats_lengths
+
+    def compute_correct(
+            self,
+            logits,
+    ):
+        if logits.numel() == 0:
+            return 0, 0
+        else:
+            assert logits.dim() > 1, logits.shape
+            max = logits.argmax(-1) == 0
+            min = logits.argmin(-1) == 0
+            both = max & min
+            corr = max.long().sum().item() - both.long().sum().item()
+            count = max.numel()
+            return corr, count
+        
+    def _calc_att_loss(
```

Comment:
It's a bit confusing to use "att"("attention") here. please rename it in this script.